### DISCUSSION TITLE
Placing Filename as text in a transparent area below image

### DISCUSSION THREAD

**Discussion by Trailboy** [ImageMagick/ImageMagick#4054] [Help] [answered]
Hi all.  

Sorry to have to ask for help again. 

I would like to use Imagemagick to convert an image such that extra transparent space is placed below the image and to have the filename inserted into that space like the image below (except with the white space as transparent):

![Jane Doe](https://user-images.githubusercontent.com/64076927/128611860-94793ecb-98ee-4731-a3a3-c20adf3688c5.JPG)

I have tried to search and experiment, but have only so far come up with placing the filename on top of the image using the code:

```
for %%F in (*.jpg) do magick %%F ^
	-resize 800x800 ^
	-gravity south ^
	-undercolor none ^
	-fill black ^
	-pointsize 30 ^
	-annotate +0+20 %%f ^
	%%~nF_Annotated.png
```

which gives me this sample image

![JaneDoe_Annotated](https://user-images.githubusercontent.com/64076927/128611660-b8e1976b-f2cf-4c2b-a217-505d7dc1d273.png)

but I still have several issues to resolve :

1. I am not not extending the canvas with a transparent area.  I have tried in a separate code the montage command from previous advice given to others, but can't get that to work.
2. Filenames with spaces are not working at all, and the command breaks.  Is it possible to do this and retain the spaces?  Unfortunately, spaces (i.e. peoples names) are necessary for my purposes.
3. I have to remove the extension .jpg, but can't figure out how to include only the filename.  
4. Longshot  -  Is it possible to pull data from the IPTC Job Identifier field rather than the filename to place under the image?  The pupil full name is actually stored in that field in all our .jpegs, and so I could avoid the problem with spaces in filenames 

Any help from you very knowledgeable people would be greatly appreciated (yet again).  

Thanks

**Comment by fmw42** [comment]
Your small image does not have the text in a transparent area. It is opaque white. Is that what you want or do you really want the white to be transparent?  

What is your ImageMagick version and platform/OS?

Post an original image without any text and with the IPTC data in it.

All of what you want should be doable!

**Reply by Trailboy** [reply]
Hi Fred.  

I would prefer it to be transparent, if that is possible.  Transparent means we can have different backgrounds and just change the text colour to suit.  

Win7 / Win 10.  Happy to update our IM version though.  

![IM-Version](https://user-images.githubusercontent.com/64076927/128612868-d066be76-e12d-450c-8813-42e5a825e4ee.JPG)

Give me a little while for the image, I will have to create a generic image with all of our normal metadata intact.  The sensitive nature of the data should be obvious to all.  

I'll reply with a link shortly.

**Reply by fmw42** [reply]
Can you just post one that does not have sensitive information or dummy data?

**Comment by snibgo** [comment]
@Trailboy: What is your platform? It seems to be Windows BAT.

You have:
```
-annotate +0+20 %%f
```
If %%f contains spaces, you need to quote it "%%f". Also quote the file names.

I suggest you should be consistent: use either %%f or %%F throughout. Don't use both.

If you want the file name only, without extension, use "%%~nf". This is a Windows facility, not IM.

You seem to want to enlarge the image, with extra space at the bottom to write the name. One method is to append two images together.

**Comment by Trailboy** [comment]
Schools live and breathe on their data nowadays.  Stripping out wasn't that easy. so started again with a generic image.  

Try this drive link for a quick generic image, but one that does have a John Doe in the relevant IPTC Job Identifier field :

https://drive.google.com/drive/folders/12o4qifV2hWMfjJgrgwM1wqcIrWwD4ngu?usp=sharing

and use the outline image.  

---

Hi Snibgo!

Correct, Windows BAT files.  We probably need to run this command over thousands of images, so I am assuming that BAT files would be most appropriate.    

Enlarge, yes.  In my (very limited) understanding, I thought this would be canvas 'extend' though.  

The original files are full res dSLR, but images created as above need to be no bigger than about 800px or so for quick access by teachers and for index prints.  

Thanks for how to remove the extension.

**Comment by snibgo** [comment]
Oops, I got that wrong, sorry. Your Windows loop variable is %F but your annotation uses the IM percent escape %f. For IM percent escapes, see https://imagemagick.org/script/escape.php . The one you want is %t.

```
for %%F in (*.jpg) do magick "%%F" ^
  -resize 800x800 ^
  ( -size 800x100 xc:None ^
    -gravity Center ^
    -pointsize 30 ^
    -annotate +0+20 "%%t" ^
  ) ^
  -append ^
  "%%~nF_Annotated.png"
```
I suspect a better annotation offset would be "+0+0".

**Reply by fmw42** [reply]
With IM 7, you might want to use label: rather than -annotate to auto fit the text to the area. I do not know how long the text will be that he wants to extract. In that case, replace xc:none with -background none and use label:"%t". Double the % as needed for .bat.

**Comment by snibgo** [comment, marked as answer]
> ... I thought this would be canvas 'extend' though.

Yes, this can be done with "-extent":
```
for %%F in (*.jpg) do magick "%%F" ^
  -resize 800x800 ^
  -background None ^
  -gravity North ^
  -extent x850 ^
  -gravity South ^
  -undercolor none ^
  -fill black ^
  -pointsize 30 ^
  -annotate +0+20 %%t ^
  "%%~nF_Annotated.png"
```
Like many tasks, there are many methods that can be used.

**Reply by Trailboy** [reply]
Thank you Snibgo, that does exactly what I require.  

Even allows for the space in the filename.  

I think at this point that filename is probably better than extracting the pupil name data directly from the IPTC field.  The IPTC field includes their full names.  Using filenames we can use a file renamer to crop filenames at the first space to have pupil images more informally (i.e. with only their first names) if required.

**Comment by Trailboy** [comment]
Thanks to all, you're a credit to this forum.

**Reply by Trailboy** [reply]
Thought I would attach a sample output image for future reference.  As a GIMP screenshot because uploaded .png doesn't show transparency chequerboard.  

Thanks again.

![JaneDoeSample](https://user-images.githubusercontent.com/64076927/128995493-597afcde-0a35-4fc6-936e-df5b7d78929e.JPG)
